File: lib.streamutils/src/main/java/de/lgohlke/streamutils/DecoratingStreamer.java
package de.lgohlke.streamutils;

import lombok.extern.slf4j.Slf4j;
import lombok.val;

import java.util.Scanner;

@Slf4j
class DecoratingStreamer {
    private final Channel<String> channel;
    private final NotifyWaiter notifyWaiter;
    private final PrefixedInputStream prefixedInputStream;

    DecoratingStreamer(PrefixedInputStream prefixedInputStream,
                       Channel<String> channel,
                       NotifyWaiter notifyWaiter) {
        this.prefixedInputStream = prefixedInputStream;
        this.channel = channel;
        this.notifyWaiter = notifyWaiter;
    }


    void capture() {
        notifyWaiter.waitForReceiverStarted();
        String prefix = prefixedInputStream.getPrefix();

        try (val scanner = new Scanner(prefixedInputStream.getStream())) {

            log.debug("scanner started: {}", prefix);
            notifyWaiter.notifySenderStarted();

            while (scanner.hasNextLine()) {
                channel.send(prefix + " " + scanner.nextLine());
            }

        } catch (Exception e) {
            log.error(e.getMessage(), e);
        }
        log.debug("scanner closed: {}", prefix);
        notifyWaiter.notifySenderStopped();
    }
}

File: lib.streamutils/src/main/kotlin/de/lgohlke/streamutils/DecoratingStreamer.kt
package de.lgohlke.streamutils

import org.slf4j.LoggerFactory
import reactor.core.CoreSubscriber
import reactor.core.publisher.Flux
import java.util.*

internal class DecoratingStreamer(private val prefixedInputStream: PrefixedInputStream) : Flux<String>() {
    companion object {
        @Suppress("JAVA_CLASS_ON_COMPANION")
        private val log = LoggerFactory.getLogger(javaClass.enclosingClass)
    }

    override fun subscribe(p0: CoreSubscriber<in String>) {
        val prefix = prefixedInputStream.prefix
        try {
            Scanner(prefixedInputStream.stream).use { scanner ->
                log.warn("scanner started: {}", prefix)
                while (scanner.hasNextLine()) {
                    val nextLine = scanner.nextLine()
                    log.warn("scanner [{}] send: {}", prefix, nextLine)
                    p0.onNext("$prefix $nextLine")
                }
                p0.onComplete()
            }
        } catch (e: Exception) {
            if (e is NullPointerException) {
                log.error("", e)
            } else {
                log.error(e.message, e)
            }
            p0.onError(e)
        }
        log.warn("scanner closed: {}", prefix)
    }
}